components/PortalStatistics.lua

```diff
@@ -0,0 +1,1407 @@
+			return false
+		elseif startsEmpty and value == true then
+			return false
+		else
+			startsEmpty = false
+			return true
+		end
+	end)
+
+	categoryNames = Array.filter(categoryNames, function(_, catIndex)
+			return Logic.readBool(isEmptyCategory[catIndex]) end)
+
+	seriesData = Array.map(seriesData, function(_, index)
+			return Array.filter(seriesData[index], function(_, catIndex)
+					return Logic.readBool(isEmptyCategory[catIndex]) end)
+		end)
+	return categoryNames, seriesData
+end
+
+
+---@param args table
+---@param params table
+---@return table
+function StatisticsPortal._getChartConfig(args, params)
+	local isForTeam = String.isNotEmpty(args.team) or Logic.readBool(args.isForTeam)
+	local customInputs = StatisticsPortal._splitOrDefault(args.customInputs, params.defaultInputs)
+	local opponentName
+	if isForTeam then
+		opponentName = args.team
+	else
+		opponentName = args.player
+	end
+
+	return {
+		processFunction = params.processFunction,
+		variable = params.variable,
+		catLabel = params.catLabel,
+		flipAxes = params.flipAxes or false,
+		axisRotate = params.axisRotate or 0,
+		emphasis = params.emphasis or 'series',
+		customInputs = customInputs,
+		customLegend = StatisticsPortal._splitOrDefault(args.customLegend, customInputs),
+		customYears = args.customYears,
+		startYear = args.startYear or Info.startYear,
+		yearBreakdown = Logic.readBool(args.yearBreakdown),
+		removeEmptyCategories = Logic.readBool(args.removeEmptyCategories),
+		removeEmptySeries = Logic.readBool(args.removeEmptySeries),
+		chartType = args.chartType or 'bar',
+		stackType = args.stackType or 'total',
+		isForTeam = isForTeam,
+		opponentName = opponentName,
+		opponentType = isForTeam and Opponent.team or Opponent.solo,
+		maxOpponents = tonumber(args.maxOpponents) or MAX_OPPONENT_LIMIT,
+		height = tonumber(args.height) or 400,
+		width = tonumber(args.width) or 1400,
+	}
+end
+
+
+---@param tablePlace number
+---@param item table
+---@param config table
+---@return table
+function StatisticsPortal._defaultProcessFunction(tablePlace, item, config)
+	local earnings
+	if String.isNotEmpty(config.opponentName) and item.opponenttype == Opponent.team then
+		earnings = config.isForTeam and item.prizemoney or item.individualprizemoney
+	else
+		earnings = item.prizemoney
+	end
+	return tablePlace + Math._round(earnings or 0, DEFAULT_ROUND_PRECISION)
+end
+
+
+---@param player table
+---@return table
+function StatisticsPortal._toOpponent(player)
+	return {
+		type = Opponent.solo,
+		players = {{
+			pageName = player.pagename,
+			displayName = player.id,
+			flag = player.nationality,
+			team = String.isNotEmpty(player.team) and player.team or nil,
+		}},
+	}
+end
+
+
+---@param input string
+---@param default table?
+---@return table
+function StatisticsPortal._splitOrDefault(input, default)
+	if String.isEmpty(input) then
+		return default --[[@as table]]
+	end
+	return Array.map(mw.text.split(input, ',', true), String.trim)
+end
```

Comment:
```suggestion
---@param input string?
---@param default table?
---@return table
function StatisticsPortal._splitOrDefault(input, default)
	if String.isEmpty(input) then
		return default --[[@as table]]
	end
	---@cast input -nil
	return Array.map(mw.text.split(input, ',', true), String.trim)
end
```
note the `--[[@as table]]` is still wrong in case of default being nil ...
in case default is nil AND input is empty this function factually returns nil

Reply:
Should the return be `---@return table?` then

Reply:
Or does the array.map return an empty table if input is nil

Reply:
the text split would error on nil, but since you do an is empty check above the input is not empty when it gets to the array (hence the casting it as not nil)

yes technically this function in its current form returns `table?`
but that would cause anno issues in other places

so the best way would probably be to make it not return nil via an ` or {}` when returning the default